nexus/reconfigurator/planning/src/blueprint_builder/builder.rs

```diff
@@ -920,6 +927,160 @@ impl<'a> BlueprintBuilder<'a> {
         Ok(EnsureMultiple::Changed { added: num_crdb_to_add, removed: 0 })
     }
 
+    pub fn sled_promote_internal_ntp_to_boundary_ntp(
+        &mut self,
+        sled_id: SledUuid,
+    ) -> Result<EnsureMultiple, Error> {
+        // The upstream NTP/DNS servers and domain _should_ come from Nexus and
+        // be modifiable by the operator, but currently can only be set at RSS.
+        // We can only promote a new boundary NTP zone by copying these settings
+        // from an existing one.
+        let (ntp_servers, dns_servers, domain) = self
+            .parent_blueprint
+            .all_omicron_zones(BlueprintZoneFilter::All)
+            .find_map(|(_, z)| match &z.zone_type {
+                BlueprintZoneType::BoundaryNtp(zone_config) => Some((
+                    zone_config.ntp_servers.clone(),
+                    zone_config.dns_servers.clone(),
+                    zone_config.domain.clone(),
+                )),
+                _ => None,
+            })
+            .ok_or(Error::NoBoundaryNtpZonesInParentBlueprint)?;
+
+        self.sled_promote_internal_ntp_to_boundary_ntp_with_config(
+            sled_id,
+            ntp_servers,
+            dns_servers,
+            domain,
+        )
+    }
+
+    pub fn sled_promote_internal_ntp_to_boundary_ntp_with_config(
+        &mut self,
+        sled_id: SledUuid,
+        ntp_servers: Vec<String>,
+        dns_servers: Vec<IpAddr>,
+        domain: Option<String>,
+    ) -> Result<EnsureMultiple, Error> {
+        // Check the sled id and return an appropriate error if it's invalid.
+        let _ = self.sled_resources(sled_id)?;
+
+        let sled_zones = self.zones.change_sled_zones(sled_id);
+
+        // Find the internal NTP zone and expunge it.
+        //
+        // TODO-cleanup Is there ever a case where we might want to do some kind
+        // of graceful shutdown of an internal NTP zone? Seems unlikely...
```

Comment:
I can't see why.  Is there something you have in mind?

Reply:
Not really. Maybe if we were doing some kind of graceful cleanup of all the zones on a sled? But even then I don't think there's anything to actually _do_. I'll remove this comment.